tests/saver/test_save_prediction.py

```diff
@@ -0,0 +1,105 @@
+import numpy as np
+import pandas as pd
+
+from scipy.stats import norm
+
+from cascade.saver.save_prediction import uncertainty_from_prediction_draws, _predicted_to_uploadable_format
+
+
+def test_uncertainty_from_prediction_draws():
+    rng = np.random.RandomState(4398221)
+
+    num_draws = 100
+    locations = [1, 2]
+    age_lower, age_upper = zip(*[(0, 0.5), (50, 60)])
+    sexes = [-0.5]
+    time_lower, time_upper = zip(*[(y, y) for y in [2000, 2020]])
+    integrands = ["prevalence", "Sincidence"]
+
+    df = pd.DataFrame(
+        index=pd.MultiIndex.from_product(
+            [locations, age_lower, age_upper, time_lower, time_upper, integrands, sexes],
+            names=["location", "age_lower", "age_upper", "time_lower", "time_upper", "integrand", "s_sex"],
+        )
+    ).reset_index()
+
+    true_mean = np.linspace(0.0001, 2.0, len(df))
+    rng.shuffle(true_mean)
+    true_std = np.linspace(0.00001, 0.001, len(df))
+    rng.shuffle(true_std)
+    df["true_mean"] = true_mean
+    df["true_std"] = true_std
+
+    draws = []
+    for sample_index in range(num_draws):
+        new_df = df.copy()
+        new_df["sample_index"] = sample_index
+        new_df["mean"] = new_df[["true_mean", "true_std"]].apply(
+            lambda r: rng.normal(r["true_mean"], r["true_std"]), axis="columns"
+        )
+        draws.append(new_df)
+
+    with_uncertainty = uncertainty_from_prediction_draws(
+        [
+            draw[
+                [
+                    "location",
+                    "age_lower",
+                    "age_upper",
+                    "time_lower",
+                    "time_upper",
+                    "integrand",
+                    "s_sex",
+                    "sample_index",
+                    "mean",
+                ]
+            ]
+            for draw in draws
+        ]
+    )
+
+    with_uncertainty.set_index(
+        ["location", "integrand", "age_lower", "age_upper", "time_lower", "time_upper", "s_sex"]
+    ).sort_index()
+    source = pd.concat(draws).drop_duplicates(
+        ["location", "integrand", "age_lower", "age_upper", "time_lower", "time_upper", "s_sex"]
+    )
+    source = source.set_index(
+        ["location", "integrand", "age_lower", "age_upper", "time_lower", "time_upper", "s_sex"]
+    ).sort_index()
+    source["lower"] = source[["true_mean", "true_std"]].apply(
+        lambda r: norm(r["true_mean"], r["true_std"]).ppf(0.025), axis="columns"
+    )
+    source["upper"] = source[["true_mean", "true_std"]].apply(
+        lambda r: norm(r["true_mean"], r["true_std"]).ppf(0.975), axis="columns"
+    )
+
+    assert np.allclose(with_uncertainty["lower"], source["lower"], rtol=0.04)
+    assert np.allclose(with_uncertainty["upper"], source["upper"], rtol=0.1)
```

Comment:
These are fine. Yes, they seem wide.

Reply:
They're wide but the numbers involved are very small. As I increased the number of samples I actually had to loosen these to accommodate the points with the lowest std. I don't really have a sense for how reasonable that is.

Reply:
Hmmm. The math looks right to me. These stats tests are difficult.